### PR TITLE
feat:improve iac test json output

### PR DESCRIPTION
**Pull request by rontalx**
#### What does this PR do?
Formats and improves the response we get from Registry's `iac-test` for use in our JSON output.
Main changes:
- Rename `cloudConfigResults` -> `infrastructureAsCodeIssues`
- Rename `cloudConfigPath` -> `path`
- Remove the following fields: `from`, `name`
- Improved names & typings.
- Reduce tons of boilerplate code in the relevant `iac test` and clean them up.

#### old structure
```
"cloudConfigResults": [
  {
      "id": "SNYK-CC-K8S-15",
      "from": [],
      "title": "LoadBalancer Service without SourceRange allows all traffic",
      "description": "...",
      "type": "k8s",
      "subType": "Service",
      "name": "LoadBalancer Service without SourceRange allows all traffic",
      "severity": "medium",
      "cloudConfigPath": [
        "[DocId: 3]",
        "service",
        "spec",
        "loadBalancerSourceRanges"
      ],
      "isIgnored": false
    }
]
```

#### new structure
```
"infrastructureAsCodeIssues": [
  {
      "id": "SNYK-CC-K8S-15",
      "title": "LoadBalancer Service without SourceRange allows all traffic",
      "description": "...",
      "type": "k8s",
      "subType": "Service",
      "severity": "medium",
      "path": [
        "[DocId: 3]",
        "service",
        "spec",
        "loadBalancerSourceRanges"
      ],
      "isIgnored": false
    }
]
```

**Note**
There's a refactor I did for the test structure. 
there's more to do, but for the sake of releasing this PR I'm pushing it to a separate task.

#### How should this be manually tested?
Running the `iac-test` flows both with and without the `--json` flag.

#### Any background context you want to provide?
Slack Discussion: https://snyk.slack.com/archives/CMM9V3XPY/p1596616315051300

#### What are the relevant tickets?
Jira Ticket: https://snyksec.atlassian.net/browse/CC-257

### CODE REVIEW

File: src/lib/snyk-test/iac-test-result.ts [hunk 1]
```diff
@@ -9,15 +9,70 @@ export interface AnnotatedIacIssue {
...
   subType: string;
+  // Legacy fields from Registry, unused.
+  name?: string;
+  from?: string[];
 }
 
-export interface IacTestResult extends BasicResultData {
+type FILTERED_OUT_FIELDS = 'cloudConfigPath' | 'name' | 'from';
+
+export interface IacTest extends BasicResultData {
```
Comment (orkamara) [inline src/lib/snyk-test/iac-test-result.ts:19]:
```suggestion
export interface IacTestResponse extends BasicResultData {
```

File: src/lib/snyk-test/iac-test-result.ts [hunk 2]
```diff
@@ -9,15 +9,70 @@ export interface AnnotatedIacIssue {
...
+
+/**
+ * The following types represent manipulations to the data structure returned from Registry's `test-iac`.
+ * These manipulations are being done prior to outputing as JSON, for renaming fields only.
+ * The types above, IacTestResult & AnnotatedIacIssue, represent how the response from Registry actually is.
+ * These were introduced in order to prevent cascading complex changes caused by changing Registry's `test-iac` response.
+ */
+
+export interface MappedIacTest extends Omit<IacTest, 'result'> {
+  result: {
```
[the remark this answers is not in the document]
Comment (rontalx) [reply src/lib/snyk-test/iac-test-result.ts:60]: Good catch, this type shouldn't have `result` at all, but it's internals on the same level as seen here:
https://github.com/snyk/snyk/pull/1309/files/b8965f2fe3f100e46ede280112c29bf1686df09e#diff-10ce357df1fefce72d2793d5710f0a15R45

File: src/lib/snyk-test/iac-test-result.ts [hunk 3]
```diff
@@ -9,15 +9,70 @@ export interface AnnotatedIacIssue {
...
+  meta: TestDepGraphMeta;
   result: {
     cloudConfigResults: AnnotatedIacIssue[];
     projectType: string;
   };
 }
+
+const IAC_ISSUES_KEY = 'infrastructureAsCodeIssues';
+
+export const mapIaCTestResult = (iacTest: IacTest) => {
```
[the remark this answers is not in the document]
Comment (rontalx) [reply src/lib/snyk-test/iac-test-result.ts:33]: `MappedIacTest`

File: test/acceptance/cli-test/cli-test.iac-k8s.utils.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
+import { getWorkspaceJSON } from '../workspace-helper';
```
Comment (orkamara) [inline test/acceptance/cli-test/cli-test.iac-k8s.utils.ts:1]: add us as a code owner

File: test/acceptance/workspaces/iac-kubernetes/test-iac-high-result.json [hunk 5]
```diff
@@ -5,6 +5,8 @@
       {
```
Comment (orkamara) [inline test/acceptance/workspaces/iac-kubernetes/test-iac-high-result.json:5]: add us as a code owner

File: test/acceptance/workspaces/iac-kubernetes/test-iac-low-result.json [hunk 6]
```diff
@@ -5,6 +5,8 @@
       {
```
Comment (orkamara) [inline test/acceptance/workspaces/iac-kubernetes/test-iac-low-result.json:5]: add us as a code owner

File: test/acceptance/workspaces/iac-kubernetes/test-iac-medium-result.json [hunk 7]
```diff
@@ -5,6 +5,8 @@
       {
           "id": "SNYK-CC-K8S-1",
           "title": "Reducing the admission of containers with dropped capabilities",
+          "name": "Reducing the admission of containers with dropped capabilities",
```
Comment (orkamara) [inline test/acceptance/workspaces/iac-kubernetes/test-iac-medium-result.json:8]: add us as a code owner

File: src/lib/snyk-test/iac-test-result.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rontalx) [reply src/lib/snyk-test/iac-test-result.ts:33]: Is there a good reason why we are not using arrow functions? or simply a code style convention?

[the remark this answers is not in the document]
Comment (orkamara) [reply src/lib/snyk-test/iac-test-result.ts]: code style convention :)

[the remark this answers is not in the document]
Comment (orkamara) [reply src/lib/snyk-test/iac-test-result.ts]: please make that explicit as part of the function

File: src/cli/commands/test/index.ts [hunk 8]
```diff
@@ -487,7 +472,7 @@ function displayResult(
 
   if (res.packageManager === 'k8sconfig') {
     return getIacDisplayedOutput(
-      (res as any) as IacTestResult,
+      (res as any) as IacTestResponse,
```
[the remark this answers is not in the document]
Comment (rontalx) [reply src/cli/commands/test/index.ts:475]: Yeah I agree it's ugly.
Our type coverage is very bad in this repo, so as the input argument `res` of this function is of type `TestResult`, if I change it to be `TestResult | IacTestResponse` it triggers a chain of type errors which I prefer not to fix as part of this PR as it's a deep rabbit hole.
So prefer keeping the current code with this double casting that typescript requires 🤕 .

image

image

File: test/acceptance/cli-test/cli-test.iac-k8s.utils.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    getWorkspaceJSON(
+      'iac-kubernetes',
+      `test-iac-${severityThreshold}-result.json`,
+    ),
+  );
+
+  iacTestJsonAssertions(t, results, expectedResults);
+};
+
+export const iacTest = async (
```
Comment (anthogez) [inline test/acceptance/cli-test/cli-test.iac-k8s.utils.ts:69]: Due of our code convention would be nice to update this from function from <const> declaration to`export async function iacTest`, thanks!

Comment (anthogez) [inline test/acceptance/cli-test/cli-test.iac-k8s.utils.ts:69]: Due of our code convention would be nice to update, this function <const> declaration to`export async function iacTest`, thanks!

File: test/acceptance/cli-test/cli-test.iac-k8s.utils.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  t.match(
+    res,
+    `Tested iac-kubernetes for known issues, found ${numOfIssues} issues`,
+    `${numOfIssues} issue`,
+  );
+  iacTestMetaAssertions(t, res);
+};
+
+export const iacTestMetaAssertions = (t, res) => {
+  const meta = res.slice(res.indexOf('Organization:')).split('\n');
```
Comment (anthogez) [inline test/acceptance/cli-test/cli-test.iac-k8s.utils.ts:93]: as I mentioned also this one to `function iacTestMetaAssertions`, thanks!

File: src/lib/snyk-test/iac-test-result.ts [hunk 11]
```diff
@@ -30,22 +31,26 @@ export interface IacTest extends BasicResultData {
 
 const IAC_ISSUES_KEY = 'infrastructureAsCodeIssues';
 
-export const mapIaCTestResult = (iacTest: IacTest) => {
+export const mapIacTestResult = function(
```
Comment (anthogez) [inline src/lib/snyk-test/iac-test-result.ts:34]: I appreciate the intention of `function declaration in a const` but would be good if you update it to

`export function mapIacTestResult`

Thanks!

File: src/lib/snyk-test/iac-test-result.ts [hunk 12]
```diff
@@ -55,22 +60,25 @@ export const mapIaCTestResult = (iacTest: IacTest) => {
...
+  projectType: string;
 }
 
 export interface MappedAnnotatedIacIssue
   extends Omit<AnnotatedIacIssue, FILTERED_OUT_FIELDS> {
   path: string[];
 }
 
-export const mapIacIssue = (
+export const mapIacIssue = function(
```
Comment (anthogez) [inline src/lib/snyk-test/iac-test-result.ts:79]: same already mentioned ☝️

File: .github/CODEOWNERS [hunk 13]
```diff
@@ -11,4 +11,8 @@ src/lib/snyk-test/iac-test-result.ts @snyk/cloudconfig
 src/lib/snyk-test/payload-schema.ts @snyk/cloudconfig
 src/lib/snyk-test/run-iac-test.ts @snyk/cloudconfig
 test/acceptance/cli-test/cli-test.iac-k8s.spec.ts @snyk/cloudconfig
```
Comment (anthogez) [inline .github/CODEOWNERS:13]: Please improve this commit message, `fix: cr comments & improve types`. 
We need to think about our users when it comes to `public release notes`, thanks!

Comment (rontalx) [reply .github/CODEOWNERS:13]: @anthogez I'm going to squash merge these commits into a single commit, so this wouldn't go into the release notes.

Comment (anthogez) [reply .github/CODEOWNERS:13]: Cool 👌

**Comment by rontalx** [comment]
@anthogez 
Regarding the arrow function comments, I'm fine with any code-style convention and have fixed your comments.
However, we should have this enforced as a lint rule - as currently across the codebase we have lots of usage with arrow functions.

**Pull request merged**